### ISSUE TITLE
[VR-SG-HAPPO] seed1 陷入低 SU 吞吐鞍点：high-order validation 与探索闭环导致辅助梯度关闭

### ISSUE THREAD

**Issue by lizhuoran1019** [lizhuoran1019/HARL#46]
# [VR-SG-HAPPO] seed1 陷入低 SU 吞吐鞍点：high-order validation 与探索闭环导致辅助梯度关闭

## 背景

在 `feat/vr-sg-happo` / `codex/issue-44-corr-lcb-validation` 相关实验中，基于 8 个 seed 的 TensorBoard 日志观察到：seed1 明显陷入低性能鞍点。该问题不是数值爆炸、KL 失控或 critic 崩溃，而是策略进入“低冲突、低 SU 吞吐”的保守局部解后，high-order auxiliary 无法提供有效逃逸梯度。

当前关键配置：

- `k_agents_per_group = 3`
- `use_high_order_advantage = true`
- `use_validated_counterfactual_high_order = true`
- `partition_selection_mode = greedy`
- `use_cover_exploration = false`
- `high_order_adv_coef_end = 0.04`
- `projection_lcb_eps = 1e-8`
- `projection_fit_split_ratio = 0.5`

## 实验现象

8 个 seed 的最终评估结果如下：

| seed | eval reward final | train reward final | eval loss rate | PU rx | SU rx | 现象 |
|---:|---:|---:|---:|---:|---:|---|
| 1 | 4665.0 | 4583.6 | 0.0000 | 311 | 311 | 低 SU 吞吐鞍点 |
| 2 | 5202.5 | 5061.8 | 0.0136 | 303 | 438 | 正常 |
| 3 | 5490.0 | 5366.0 | 0.0000 | 302 | 494 | 正常 |
| 4 | 5596.0 | 5375.9 | 0.0065 | 310 | 501 | 最好 |
| 5 | 5208.5 | 5037.3 | 0.0422 | 299 | 455 | loss 偏高但吞吐正常 |
| 6 | 5505.0 | 5381.2 | 0.0000 | 309 | 483 | 正常 |
| 7 | 5365.0 | 5322.1 | 0.0000 | 306 | 461 | 正常 |
| 8 | 5101.5 | 5335.8 | 0.0215 | 489 | 47 | PU/SU 极端失衡 |

seed1 的主要问题不是丢包，而是 SU 吞吐长期偏低：

- seed1 `eval loss_rate = 0`
- seed1 `PU rx = 311`
- seed1 `SU rx = 311`
- 正常高分 seed 的 SU rx 通常在 461–501 区间

说明 seed1 学到的是“安全但低吞吐”的保守协议，而不是通信失败协议。

## 日志诊断

seed1 后期 high-order 相关指标显示 auxiliary 基本被关闭：

- `return_high_order/accepted_fraction` 最后为 0
- `high_order/effective_adv_coef` 最后为 0
- `return_high_order/reject_non_positive_lcb_fraction` 最后为 1
- `return_high_order/projection_lcb_mean` 最后约为 -0.594
- `return_high_order/val_alignment_mean` 最后约为 -0.185

同时 PPO 主更新没有失控：

- `group/kl` 很低
- `group/ratio_std` 很小
- `group/kl_early_stop = 0`
- 未观察到梯度爆炸或 KL early stop 导致的训练中断

因此问题不是优化不稳定，而是有效逃逸方向缺失。

## 根因分析

当前训练闭环可以概括为：

```text
策略进入保守局部解
→ on-policy 数据缺少“提升 SU 且不伤 PU”的正高阶样本
→ return-validated projection 的 LCB ≤ 0
→ high-order auxiliary 被 hard gate 关闭
→ actor 只剩 base PPO 小步更新
→ 策略继续保守
→ 下一轮仍采不到正高阶样本
```

数学上，当前组级目标近似为：

```text
L_G = L_base^G + alpha * 1{LCB_G > eps} * L_high_order^G
```

当 seed1 进入鞍点后：

```text
∇ L_base^G ≈ 0 或方向偏保守
```

而由于：

```text
LCB_G <= eps
```

实际变成：

```text
∇ L_G ≈ ∇ L_base^G
```

即 high-order 逃逸梯度被关掉。

### 关键原因 1：greedy partition 缺少系统覆盖

当前 `partition_selection_mode = greedy`，并且 `use_cover_exploration = false`。这会导致 partition 更偏向当前已有高分组合，而不是主动覆盖“证据不足但可能有用”的组合。

改为 `hybrid` 可以缓解，但不根治。因为 hybrid 只是增加随机性，不能保证关键 group 和关键 joint action pattern 被系统采样。

### 关键原因 2：validation 使用 batch-local hard gate

当前 return-validated projection 在当前 batch 上拟合并验证。如果当前 batch 来自保守策略，样本中本来就缺少正高阶协同，LCB 很容易非正。

一旦 `LCB <= eps`，该 group 的 high-order auxiliary 直接被置零。这个机制会把“当前证据不足”误处理为“该方向无用”。

### 关键原因 3：没有跨 rollout 的 high-order 证据记忆

当前没有长期记录每个 group：

- 被采样多少次
- joint action pattern 覆盖多少
- 历史 alignment 均值和方差
- 长期 LCB 是否逐渐转正

因此每个 batch 都像重新考试。seed1 一旦进入保守策略，单 batch 证据长期不足，high-order 方向就长期无法恢复。

### 关键原因 4：k=3 的协同覆盖上限

当前 `k_agents_per_group=3`，单个组最多直接感知三阶以内组内协同。如果实际 MAC 改善需要更大范围的联合避让、错峰接入或 4 个以上 agent 的协同，k=3 会增加 seed 敏感性。

这不是当前 seed1 的唯一原因，但会放大“采样不足 + validation 过硬”的问题。

## 为什么只改 hybrid 不够

`hybrid` 的作用是让 partition 偶尔选择非 greedy 组合。它能增加一定探索，但不能解决：

1. group 是否被系统覆盖；
2. group 内 joint action pattern 是否被充分覆盖；
3. 当前 batch LCB 非正时 auxiliary 是否直接关闭；
4. 历史证据是否能跨 rollout 累积；
5. high-order 方向是否能在安全约束下以置信度方式参与更新。

因此，`greedy -> hybrid` 应作为消融实验，但不是根治方案。

### 方案总览

将当前训练闭环从：

```text
greedy partition
→ batch-local validation
→ hard accepted/rejected
→ auxiliary on/off
```

改为：

```text
coverage-aware partition
→ high-order support memory
→ global confidence estimation
→ confidence-weighted constrained auxiliary update
```

核心目标是：先系统性地产生和记录高阶协同样本，再根据长期证据决定 auxiliary 的使用强度。

---

## 修改 1：新增 HighOrderSupportMemory

新增长期记忆模块，记录每个 group 的高阶支持度与历史效果。

建议维护：

```python
class HighOrderSupportMemory:
    def update(self, group, joint_actions, base_adv, high_order_adv, active_mask):
        ...

    def coverage(self, group):
        ...

    def effective_sample_count(self, group):
        ...

    def alignment_mean(self, group):
        ...

    def alignment_var(self, group):
        ...

    def global_lcb(self, group):
        ...

    def uncertainty_bonus(self, group):
        ...
```

每个 group 维护：

```text
visit_count[G]
joint_action_support[G]
coverage[G]
alignment_mean[G]
alignment_var[G]
effective_sample_count[G]
global_lcb[G]
```

其中：

```text
alignment = base_adv * high_order_adv
```

长期 LCB：

```text
global_lcb[G] = alignment_mean[G] - c * sqrt(alignment_var[G]) / sqrt(N_eff[G])
```

### 新增日志

```text
group_support/coverage_mean
group_support/coverage_min
group_support/effective_sample_count_mean
group_support/global_lcb_mean
group_support/global_lcb_min
group_support/uncertainty_bonus_mean
```

---

## 修改 2：partition 改为 coverage-aware，而不是纯 greedy/hybrid

当前 partition score 主要来自当前 interaction score。建议改为：

```text
score_G =
    interaction_score_G
  + beta_ucb * uncertainty_bonus_G
  + beta_cov * (1 - coverage_G)
```

其中：

- `interaction_score_G`：当前已有交互强度
- `uncertainty_bonus_G`：样本不足或方差较大时的探索奖励
- `1 - coverage_G`：joint action pattern 覆盖不足奖励

新增配置：

```json
{
  "partition_selection_mode": "coverage_ucb",
  "coverage_weight_start": 0.5,
  "coverage_weight_end": 0.05,
  "uncertainty_weight_start": 0.3,
  "uncertainty_weight_end": 0.03,
  "coverage_min_quota": 0.6,
  "coverage_quota_window": 20
}
```

### 原则

早期优先覆盖，后期逐渐转向 exploitation。

这不是随机探索，而是有目标地补齐高阶协同样本。

---

## 修改 3：validation 从 batch-local LCB 改为 global LCB

当前逻辑类似：

```python
accepted = batch_lcb > projection_lcb_eps
```

建议改为：

```python
global_lcb = support_memory.global_lcb(group)
confidence = sigmoid(global_lcb / tau)
```

保留 hard reject，但只用于结构性无效情况：

```text
rank_deficient -> reject
insufficient_samples -> collect_more
global_lcb very negative -> reject
weak evidence -> low confidence, not zero
positive global_lcb -> high confidence
```

新增配置：

```json
{
  "high_order_confidence_temperature": 0.03,
  "global_lcb_min_samples": 128,
  "global_lcb_decay": 0.95,
  "hard_reject_lcb_threshold": -0.5
}
```

---

## 修改 4：actor loss 从 hard gate 改为 confidence-weighted constrained auxiliary

当前形式：

```text
loss = base_loss + accepted * aux_coef * high_order_loss
```

建议改为：

```text
loss =
    base_ppo_loss
  + confidence_G * high_order_aux_coef * high_order_loss
  + safety_penalty
```

安全约束用于避免 high-order 方向破坏 base return：

```text
safety_penalty = lambda_safe * relu(-base_improvement_lcb - delta)
```

含义：

- high-order 方向可以在证据弱时小幅参与；
- 证据强时增强；
- 若明显伤害 base return，则压制；
- 不再因为单 batch LCB 非正就直接归零。

新增日志：

```text
high_order/confidence_mean
high_order/confidence_min
high_order/global_lcb_mean
high_order/safety_penalty
high_order/effective_aux_coef_global
```

---

## 修改 5：训练日程分阶段

建议采用三阶段训练：

### Stage 1：coverage-first warmup

训练前 30%：

```text
coverage_weight 高
uncertainty_weight 高
high_order_aux_coef 低
目标：采够高阶支持样本
```

### Stage 2：confidence-weighted high-order training

训练 30%–70%：

```text
coverage_weight 衰减
global_lcb 开始主导 confidence
high_order_aux_coef 提升
目标：利用已验证的高阶方向
```

### Stage 3：validated exploitation

训练 70%–100%：

```text
coverage_weight 低
confidence 高的 group 主导更新
目标：稳定提升吞吐与公平性
```

---

## 消融实验设计

建议按以下顺序验证：

| 实验 | 改动 | 目的 |
|---|---|---|
| A | 当前 greedy 版本 | baseline |
| B | `partition_selection_mode=hybrid` | 验证随机换组是否缓解 |
| C | coverage-aware partition | 验证系统覆盖是否有效 |
| D | C + HighOrderSupportMemory | 验证长期记忆是否减少误拒绝 |
| E | D + global LCB confidence | 验证 batch-local hard gate 是否是主因 |
| F | E + constrained auxiliary loss | 验证最终根治效果 |

## 验收标准

根治成功的标准不只看 reward，还要看闭环是否被打破。

关键指标：

```text
eval_metrics/su_total_e2e_rx_count
eval_metrics/pu_total_e2e_rx_count
return_high_order/accepted_fraction
return_high_order/reject_non_positive_lcb_fraction
high_order/effective_aux_coef
group_support/coverage_mean
group_support/global_lcb_mean
high_order/confidence_mean
partition/history/group_coverage_fraction
```

期望表现：

- seed1 的 SU rx 从约 311 提升到 450+
- `return_high_order/accepted_fraction` 不再长期为 0
- `reject_non_positive_lcb_fraction` 明显下降
- `high_order/effective_aux_coef` 长期非零
- `group_support/coverage_mean` 上升
- `global_lcb_mean` 逐步转正
- reward 均值提升，seed 间方差下降

## 预期收益

该修改不是调阈值，而是改变 high-order learning 的闭环：

当前：

```text
没证据 -> 不探索 -> 更没证据
```

修改后：

```text
主动覆盖 -> 累积证据 -> 置信更新 -> 稳定利用
```

预期可以解决 seed1 低 SU 吞吐鞍点问题，并降低不同 seed 对初始探索轨迹的敏感性。

## 任务拆分

- [ ] 新增 `HighOrderSupportMemory`
- [ ] 添加 group coverage / support / global LCB 日志
- [ ] 实现 `coverage_ucb` partition selection mode
- [ ] 将 return validation 从 batch-local LCB 扩展为 global LCB confidence
- [ ] 将 high-order auxiliary 从 hard gate 改为 confidence-weighted update
- [ ] 添加 base improvement safety penalty
- [ ] 跑 A–F 消融实验
- [ ] 对比 seed1 是否脱离低 SU 吞吐鞍点
- [ ] 检查是否降低 8 seed 方差

## 结论

seed1 的核心问题不是数值不稳定，也不是单纯 greedy 选组，而是当前 high-order 机制把“探索、验证、利用”全部绑定在同一个 on-policy batch 上。

当策略进入保守鞍点后，batch 内自然缺少正高阶协同样本，LCB 非正，auxiliary 被关闭，策略继续保守，形成闭环。

根治方案是引入：

```text
HighOrderSupportMemory
+ coverage-aware partition
+ global LCB confidence
+ constrained auxiliary update
```

使算法先系统性采样高阶协同，再用长期证据决定如何利用，而不是让单批次 validation 直接关闭高阶学习。

**Comment by lizhuoran1019** [comment]
#48 已经以 squash merge 合入 `feat/vr-sg-happo`：

- merge commit: `a756e6d8e8719d49e07052afd564a7766211009f`
- 合入内容定位为 infrastructure：`coverage_ucb + HighOrderSupportMemory + diagnostics`
- 默认路径不启用 actor-side `support-memory confidence gate`
- 默认不启用 joint-action Q critic 主裁判
- 默认不启用 weak cover exploration
- 不实现 `mass_control`，也不修改 actor loss 形式

因此 #48 只是为 #46 铺好了 selection-side 的覆盖记忆与诊断基础，不能视为已经解决本 issue。

#46 继续保持 open。后续验收仍应以训练结果为准，重点看：

- `high_order/effective_injection_mass` 是否恢复到有效量级；
- reward / PU、SU throughput 是否稳定改善；
- PU≈300 / SU≈300 的坏 seed 数是否下降；
- evidence-weighted high-order auxiliary allocator、injection mass control、gradient safety 等后续 PR 是否完成。

下一步建议单独开新 PR 处理 actor-side high-order auxiliary allocation，而不是继续扩大 #48 的 scope。
